WIP wiki support for microsite templates

felipemontoya opened the pull request:
This PR would allow the templates being rendered with the django_template engine, most notably the wiki, to support the templates override used for microsites. Specially the header and footer templates.

The way this PR addresses this issue is by creating a template tag that wrapps the microsite.get_template_path function used for finding the templates. Then it uses this  templatetag inside the main_django.html template.

edx-webhook commented:
Thanks for the pull request, @felipemontoya! I've created [OSPR-949](https://openedx.atlassian.net/browse/OSPR-949) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

felipemontoya commented:
Hi @chrisndodge, some of my work this week led me to support the microsites template overrides on the wiki pages, so I put together this PR to discuss if this is upstream material.

chrisndodge commented:
Thanks @felipemontoya! I'll try this out. FYI, @singingwolfboy

chrisndodge commented on lms/templates/main_django.html:
```diff
@@ -33,19 +33,17 @@
         {% include "navigation-edx.html" %}
       {% else %}
-        {% include "navigation.html" %}
+        {% include "navigation.html"|microsite_template_path %}
```
nifty trick! I didn't know this was possible.

chrisndodge commented on common/djangoapps/microsite_configuration/templatetags/microsite.py:
```diff
@@ -63,3 +63,14 @@ def microsite_css_overrides_file():
+    it is stripped before returning.
+    """
+    template_name = microsite.get_template_path(template_name)
+    return template_name[1:] if template_name[0] == '/' else template_name
```
any risk if template_name is None? That probably wouldn't happen in practice, but it seems like we might want a quick exit at the top of the function

felipemontoya replied:
I think there is very little chances of that happening since it is used as a filter on a static string. But since the consequences of that would be annoying (500 error) we might as well prevent it.

felipemontoya replied:
Now there are only 2 known use cases of this tag: "navigation.html" and "footer.html", but eventually it could be used on variables as well and the risk would rise.

chrisndodge replied:
hey, just checking back into this PR, are you going to be able to put in those None guards? Thanks. In the meantime, I'm going to pull this down and try locally.

felipemontoya replied:
Hi. I loaded up this branch on its devstack and tried to pass None in the template name. With or without the None guards it will throw an error, the difference would be if is either TemplateDoesNotExist: No exception supplied, or IndexError: string index out of range. I think it does not make a big difference, but if you think they are still required, just ping me here, and I will add them and push them right away.

chrisndodge commented:
Also, I don't think this branch is still active in your repository. I'm trying to pull from your repo to run this locally - to give it a whirl - but I'm not seeing it in: https://github.com/eduNEXT/edunext-platform

Am I looking in the right place?

chrisndodge commented:
I guess I could simply generate a .patch file from the PR and apply it locally...

felipemontoya commented:
Hi @chrisndodge, that is the repo we use for deployment, all the changes we submit to uptream are at the regular edx-platform repo. Here is the correct branch: https://github.com/eduNEXT/edx-platform/tree/fmo/wiki-support-for-microsites

chrisndodge commented:
@felipemontoya thanks, got it! Giving it a spin right now...

chrisndodge commented:
@felipemontoya hrm, I'm having trouble running that branch in my devstack. My devstack environment is close to upstream edX master, so it's likely an environment issue. I'll either need to provision a clean devstack or simple just apply the diff patch to my localdev. I'll probably do the latter to save some time and diskspace.

chrisndodge commented:
@felipemontoya thanks. Still trying to get your PR running, I'm having some devstack issues...

chrisndodge commented:
@felipemontoya OK, I'm seeing one problem. When I go to the wiki on one of our white label sites, I'm getting a template not found error. however I do have a <microsite_key>/templates/footer.html file defined (but it is a Mako template). I'll investigate, but do you have any tips?

felipemontoya commented:
The fact that it is a mako template should not make any difference. The footer I am currently using is also a mako template. The only thing I have seen here are rendering errors when the file does not have unix endings, but I don't think that is the case here. 

Now, is the path literally `/templates/footer.html`, because the templatetag strips the leading / since the include tag would not take absolute paths.

chrisndodge commented:
thanks. still having trouble here. I do have a footer.html in my microsite, but I don't have a navigation.html (not overridden). The navigation.html loads fine from the base edx-platform, but footer.html fails.

I hate to throw out the Django1.8 card, but I'm running this on 1.8, are you? I'm not sure if there might be some template loader differences between the versions.

felipemontoya commented:
Django 1.8 might actually be the cause. I have this branch in an old devstack from django 1.4. I will launch a sandbox on a new devstack running from master and see if it works. It will probably take me a couple of hours.

chrisndodge commented:
thanks, sorry for the hassle. I'm not even sure it'll make a difference...

felipemontoya commented:
Hi @chrisndodge, I found the same issue on a new devstack running django 1.8. It has to do with the way the lms/startup.py:enable_microsite function works, but I'm still looking into it. I will update the PR with both a fix and the rebase to master when I make it work.

felipemontoya commented:
@chrisndodge, basically, the MakoLoader in django1.8 has an engine which gets initialized with the TEMPLATE['DIRS'] before we had the chance to enable the microsites feature. I moved the enable_microsites to make it add the microsites_root_path before that happens.

chrisndodge commented:
@felipemontoya thanks!

felipemontoya commented:
The test env clearly didn't like the change, now I'm wondering if that is a missing configuration on the test env. Did you tried the change in your devstack to see if it works for you?

chrisndodge commented:
I should be able to try locally in a few hours. Thanks!

chrisndodge commented:
Hi @felipemontoya just swung back here. I see a lot of failures. Any ideas?

chrisndodge commented:
Hi @felipemontoya just checking in on this. Thanks.

felipemontoya commented:
hi @chrisndodge,  I have not yet worked on this to figure it out. I have been fencing deadlines for the last 2 weeks. I should be able to get some work done after this friday

chrisndodge commented:
Hi @felipemontoya I hate to tell you, but if I apply these diffs locally off of master, I'm getting an exception loading the homepage (removing the partner name from the exception msg):

TemplateLookupException at /
Cant locate template for uri '/{redact}/templates/header.html'

Also, main_django.html has changed a lot since your PR, I presume that was Comprehensive Theming's doing.

felipemontoya commented:
@chrisndodge, got it. Then this definitely needs more work. I'll get to it, once I finish the work I'm currently on. This looks related to a relative vs absolute path. Just to be sure, on the configuration you use, the microsite_root directory is not root, correct?

chrisndodge commented:
Thanks @felipemontoya

chrisndodge commented:
Just tagging @mattdrayer and @douglashall regarding open PR's involving microsites/white labels. This is an Open Source contribution.

mattdrayer commented:
@felipemontoya @chrisndodge -- is this PR in any way dependent on #10158 ?

chrisndodge commented:
@mattdrayer this is independent. However this PR was proposed before the Django 1.8 upgrade and now I can't get it to work on my 1.8 devstack.

felipemontoya commented:
Hi @mattdrayer, this is completely independent. For me it works on destack, but the change in startup.py breaks the tests. I'll take a look at this, but I am on vacation already so I'll look at this at night.

felipemontoya commented:
hi @mattdrayer , @chrisndodge. I rebased this yesterday and fixed the startup issues. The problems was that the regular django_template finders required the directory to be added before startup and mako required it after.

chrisndodge commented:
Thanks @felipemontoya. I should have a moment to pull this down and try it out today.

chrisndodge commented:
@felipemontoya @mattdrayer I pulled this down tonight and applied it locally. In general, it seems to work, but there seems to need some changes in edX's production microsites for this to fully work:

1) We need to add a ## mako statement at the top of our header.html file defined in our microsites
2) The .css override file needs to be added to the Wiki page (otherwise styling is kinda broken)

I'm going to have to dig in a bit here to see how we can pull this in without causing visual regressions in our white labels.

mattdrayer commented:
@felipemontoya let us know if you need a hand with this PR -- we have a ticket in our backlog (WL-247) reserved for providing support.

felipemontoya commented:
Hi @mattdrayer and @chrisndodge. I has out for the holidays, but I'm back at the office now. 

Obviously I have no access to the repository where you keep your microsites, but I could easily add the css override to the django_main, the same way we did it at [our repo](https://github.com/eduNEXT/edunext-platform/blob/master/lms/templates/main_django.html#L24). We did this so long ago that I did not even notice that it wasn't in the upstream repo. Now, since Chris mentioned the visual regressions in your white labels I will wait a word from you before I do that.

P.S: I think I'm not allowed to see WL-247.

asadiqbal08 commented on lms/templates/main_django.html:
```diff
@@ -28,14 +28,14 @@
     <a class="nav-skip" href="{% block nav_skip %}#content{% endblock %}">{% trans "Skip to main content" %}</a>
     {% with course=request.course %}
-      {% include "header.html" %}
+      {% include "header.html"|microsite_template_path %}
```
@mattdrayer @chrisndodge I investigated it on my side after pulling down this branch. It seems me, wiki page is based on Django templates rather than mako and after running the lms on this branch , It seems me header from microsite is not loading in wiki. You can see that edx-microsite/mit-prof-ed/templates/header.html is based on mako
wiki page is trying to load this microsite header.html

mattdrayer replied:
@felipemontoya if you can address this issue today, that would be great.

felipemontoya replied:
I am not sure I understood what you mean, but the renderer is actually capable of including Mako templates from within a django_template. That is what the `## mako` first line does. It helps the template engine to [recognize mako templates](https://github.com/edx/edx-platform/blob/master/common/djangoapps/edxmako/makoloader.py#L44) to load them with the right library.

asadiqbal08 replied:
then It means, our microsite templates should have `## mako` in first line.

felipemontoya replied:
yes, although not every single template needs it. It is only used when loading mako templates from inside django_templates. Some examples that do need it are header.html and footer.html

mattdrayer replied:
Okay, so I guess I'm not clear on what needs to happen next for this PR -- @asadiqbal08, do we need to update additional microsite templates in our repo with the mako tag prior to merging in order to avoid breakages?

asadiqbal08 replied:
@mattdrayer make tag `## mako` is missing in our microsite repo that is why there is a breakage in header section of wiki. I think we need to add up this in order to header loading. 

@mattdrayer @felipemontoya  It seems me that microsite css is not rendering in wiki. we need to override the microsite css.  I just tested it after adding the `{% microsite_css_overrides_file %}` in `main_django.html` and styles are rendering OK.

mattdrayer replied:
@asadiqbal08 -- okay, let's go ahead and make the necessary changes in the edx-microsites repository.  Please log a WL ticket to track the job and add it to the current sprint.

felipemontoya commented:
That and also tell me if you also want to have me add the CSS overrides file. I would say yes to that, but you need to test it along with your microsites repo to see that the styles are rendering OK.

mtyaka commented:
Thanks for working on this, it would be great to get microsites working correctly with django templates. I opened a very similar PR some time ago (before comprehensive theming landed).

@singingwolfboy suggested to use a custom django template loader rather than filters, see: https://github.com/edx/edx-platform/pull/9282#discussion_r37007674
I really like that idea, especially because having a dynamic django loader similar to the mako loader would also enable comprehensive theming to work on the wiki. Currently it doesn't work due to the same problem you ran into while working on this PR - when comprehensive theme startup code runs and adds new template paths to the `TEMPLATES['DIRS']`, default django template engine is already initialized and adding new folders to the `'DIRS'` entry has no effect.

felipemontoya commented:
@mattdrayer and @asadiqbal08 I just updated the main_django file adding the css overrides. Which use to be there, but where removed at [this](https://github.com/edx/edx-platform/commit/58daa365cd18a30f37a2567a3edb405ad07a5248#diff-e4a6f4d628ef0702cd6f5c7e9e7a8a5a) point.

After I did that, I went back to the comp_theming [readme file](https://github.com/edx/edx-platform/blob/master/themes/README.rst#microsites) stating that you should now place this in the `head-extra.html` which currently does not exist in the main repository. There is however a `header_extra.html` file which is empty and has a note saying it is meant to be overriden by the templates in the microsite.

Putting everything togher I am now sure, that the css overrides should not longer be there (I'll remove it next thing), but instead, you also have to place `## mako` tags on the `head-extra.html` files in your microsites to make sure they render correctly and in turn they will add the css_overrides_file.

@singingwolfboy could you please correct me if I'm wrong in the last paragraph or on this PR's code?

singingwolfboy commented:
@felipemontoya I believe you are correct. I've put in a PR to remove the unused `header_extra.html` file: https://github.com/edx/edx-platform/pull/11165

felipemontoya commented:
I tested my assumptions locally and it work, it required also that the `head-extra.html` import was filtered with the `microsite_template_path` tag. I updated the code accordingly

Again, the `head-extra.html` file in your microsite repo also needs the `## mako` tag.

mattdrayer commented:
Okay, we've merged @asadiqbal08's mako tag update to our microsites repository, so I think we're good to go there.  @felipemontoya @singingwolfboy @asadiqbal08 @mtyaka if there are no other changes you feel are necessary for this PR after having a chance to review the changeset I think we are good to move forward -- let me know.

singingwolfboy commented:
Looks fine to me. :+1:

mattdrayer commented:
@felipemontoya I think the only other thing I would like to see here is a handful of unit tests which cover the new djangoapp operations -- would that be possible to include in this PR?

felipemontoya commented:
@mattdrayer a pair of tests for the `microsite_template_path` filter should be easy. I'll add them in a while.

mattdrayer commented:
Hi @felipemontoya -- just pinging you re: the aforementioned tests -- once we have those I think this PR is good to go.

felipemontoya commented:
@mattdrayer, I added one simple test for the filter, to make sure those lines are being run during the test process. Do you reckon more complicated tests are in order there?

mattdrayer commented:
jenkins run python

mattdrayer commented:
@felipemontoya -- thanks for adding the test -- I'm still a little fuzzy on exactly what is being covered here semantically, but Jenkins is happy and the diff coverage report is showing 100% so I guess in cold world of robots and LOCs we are looking good.  If there are no other changes to be made then I'd say it's time to squash, verify one last time, and merge :+1:

felipemontoya commented:
jenkins run bokchoy

felipemontoya commented:
@mattdrayer, squashed and tested. From my end we are ready.

mattdrayer commented:
Alright, let's do it!  Thanks very much for this contribution, @felipemontoya -- merging now!
